### PR TITLE
Allow commas in CSV exporter fields

### LINKED ISSUE #1416: CSV export is malformed when any value contains a comma
# Bug Report

## Expected Behavior

When exporting logs as a CSV, all data should be well-formed.

## Actual Behavior

When any field value contains a comma, the CSV is malformed.

## Steps to Reproduce the Problem

1. Set your user display name to one that contains a comma, for example "Last Name, First Name"
2. Perform an action that gets logged in the audit log
3. Visit the audit log screen and download a CSV from the Record Actions menu at the bottom
4. Open the CSV in a spreadsheet application which will show the data in columns
5. Observe that the CSV output for the affected row is malformed due to the unquoted comma

## Suggested Fix

The `Exporter_CSV::output_file()` method should use `fputcsv()` instead of manually concatenating fields with commas.

## System Information

- Stream plugin version: 3.9.2
- WordPress version: 6.1
- PHP version: 8.0

### PR DESCRIPTION
**Pull request by tharsheblows**
Fixes #1416 

Use `fputcsv()` to format csv and write to file. 

# Checklist

- [ ] Project documentation has been updated to reflect the changes in this pull request, if applicable.
- [x] I have tested the changes in the local development environment (see `contributing.md`).
- [x] I have ~added~ updated phpunit tests.

## Release Changelog

- Fix: commas no longer break the csv exporter

### CODE REVIEW

File: exporters/class-exporter-csv.php [hunk 1]
```diff
@@ -38,12 +38,18 @@ public function output_file( $data, $columns ) {
...
+		$csv = fopen( 'php://output', 'w' );
+		fputcsv( $csv, array_values( $columns ) );
+
 		foreach ( $data as $row ) {
-			$output .= join( ',', $row ) . "\n";
+			fputcsv( $csv, $row );
 		}
+		fclose( $csv ); // phpcs:ignore WordPress.WP.AlternativeFunctions.file_system_operations_fclose
+
+		echo ob_get_clean(); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
```
Comment (tharsheblows) [inline exporters/class-exporter-csv.php:51]: @delawski I'm worried about the size of the output buffer, do you think I need to be?

Comment (tharsheblows) [reply exporters/class-exporter-csv.php:51]: See https://github.com/xwp/stream/issues/1476 and https://github.com/xwp/stream/issues/1503 for some large log sizes

Comment (delawski) [reply exporters/class-exporter-csv.php:51]: That's a good question. I guess that with a high enough number of records we may hit a PHP timeout limit (depending on the server configuration). However, some pro/enterprise users may still want to export everything.

I think a reasonable short-term solution to that issue would be to allow setting the number of records to export (as suggested [here](https://github.com/xwp/stream/issues/1476#issuecomment-2034374922)). We could default to 10000 records but make that value filterable.

A more sophisticated solution (maybe for the future) would be to rely on some async queue (Action Scheduler?) that would export the records to a file stored in the filesystem. Once everything is exported, a link to that file would be available in WP admin.

What are your thoughts?

**Review by delawski — APPROVED** [review]

File: exporters/class-exporter-csv.php (hunk 1 above)
Comment (tharsheblows) [reply exporters/class-exporter-csv.php:51]: > I think a reasonable short-term solution to that issue would be to allow setting the number of records to export (as suggested https://github.com/xwp/stream/issues/1476#issuecomment-2034374922). We could default to 10000 records but make that value filterable.

I'm not sure how to handle pagination though? I guess a first step is allowing that to be filterable and then figuring out if / where people are hitting the php limits (and testing that ourselves!)

> A more sophisticated solution (maybe for the future) would be to rely on some async queue (Action Scheduler?) that would export the records to a file stored in the filesystem. Once everything is exported, a link to that file would be available in WP admin.

I like this, it should be straightforward although we'd need to think about security but we could also make it be deleted after x amount of time too. (I 💖 Action Scheduler)

Comment (tharsheblows) [reply exporters/class-exporter-csv.php:51]: @delawski Yes, that makes sense. There's an [open issue](https://github.com/xwp/stream/issues/1476) for filtering the number. The WP CLI command has a `records_per_page` field that can be used to get an arbitrary number of records.

**Pull request merged**